pkg/vmprovider/providers/vsphere/resources/vm.go

```diff
@@ -138,42 +138,64 @@ func (vm *VirtualMachine) UniqueID(ctx context.Context) (string, error) {
 	return vm.ReferenceValue(), nil
 }
 
-func (vm *VirtualMachine) SetPowerState(ctx context.Context, desiredPowerState vmopv1.VirtualMachinePowerState) error {
-	vm.logger.V(5).Info("SetPowerState", "desiredState", desiredPowerState)
+func (vm *VirtualMachine) SetPowerState(
+	ctx context.Context,
+	existingPowerState,
+	desiredPowerState vmopv1.VirtualMachinePowerState) error {
+
+	vm.logger.V(5).Info(
+		"SetPowerState",
+		"existingState", existingPowerState,
+		"desiredState", desiredPowerState)
 
-	var powerTask *object.Task
 	var err error
+	var powerTask *object.Task
 
 	switch desiredPowerState {
 	case vmopv1.VirtualMachinePoweredOn:
 		powerTask, err = vm.vcVirtualMachine.PowerOn(ctx)
 	case vmopv1.VirtualMachinePoweredOff:
 		powerTask, err = vm.vcVirtualMachine.PowerOff(ctx)
+	case vmopv1.VirtualMachineSuspended:
+		powerTask, err = vm.vcVirtualMachine.Suspend(ctx)
+	case vmopv1.VirtualMachineGuestPoweredOff:
+		err = vm.vcVirtualMachine.ShutdownGuest(ctx)
 	default:
 		err = fmt.Errorf("invalid desired power state %s", desiredPowerState)
 	}
 
 	if err != nil {
-		vm.logger.Error(err, "Failed to change VM power state", "desiredState", desiredPowerState)
+		vm.logger.Error(err, "Failed to change VM power state",
+			"existingState", existingPowerState,
+			"desiredState", desiredPowerState)
 		return err
 	}
 
-	if taskInfo, err := powerTask.WaitForResult(ctx, nil); err != nil {
-		if te, ok := err.(task.Error); ok {
-			// Ignore error if VM was already in desired state.
-			if ips, ok := te.Fault().(*types.InvalidPowerStateFault); ok && ips.ExistingState == ips.RequestedState {
-				return nil
-			}
-		}
-		if taskInfo != nil {
-			err = errors.Wrapf(err, "%s failed", taskInfo.Name)
-		}
+	// The ShutdownGuest method does not return a task.
+	if powerTask == nil {
+		return nil
+	}
 
-		vm.logger.Error(err, "VM change power state task failed", "state", desiredPowerState)
-		return err
+	taskInfo, err := powerTask.WaitForResult(ctx, nil)
+	if err == nil {
+		return nil
 	}
 
-	return nil
+	if te, ok := err.(task.Error); ok {
+		// Ignore error if VM was already in desired state.
+		if ips, ok := te.Fault().(*types.InvalidPowerStateFault); ok && ips.ExistingState == ips.RequestedState {
```

Comment:
Maybe? I did not attempt to optimize this function, but it's a fair point. Let me take a look.

Reply:
Ah, and the answer is no, we should not. After looking at it again, remember that we only know the power state when we read it -- it could have changed. The best thing to do is to take action and handle the result. This ensures we attempt to implement the desired state regardless of what the observed state was at one time.

Reply:
Ah.  Okay.  I did not consider that the state of the VM could have changed on vCenter out-of-band of VM operator (though unsupported).  Good point.